File: notebooksc++/01-is-unique.cpp
#include <iostream>
#include <string>
#include <unordered_set>
#include <cassert>

using namespace std;

// Solution 1: Using unordered_set - O(n) time, O(n) space
bool isUniqueSet(const string& astr) {
    unordered_set<char> seen;
    for (char c : astr) {
        if (seen.count(c) > 0) {
            return false;
        }
        seen.insert(c);
    }
    return true;
}

// Solution 2: Using bit mask - O(n) time, O(1) space (for lowercase a-z only)
bool isUniqueBit(const string& astr) {
    int mark = 0;
    for (char c : astr) {
        int move_bit = c - 'a';
        if ((mark & (1 << move_bit)) != 0) {
            return false;
        }
        mark |= (1 << move_bit);
    }
    return true;
}

int main() {
    // Test isUniqueSet
    assert(isUniqueSet("abc") == true);
    assert(isUniqueSet("leetcode") == false);
    assert(isUniqueSet("algorithm") == true);
    assert(isUniqueSet("hello") == false);
    assert(isUniqueSet("") == true);

    // Test isUniqueBit
    assert(isUniqueBit("abc") == true);
    assert(isUniqueBit("leetcode") == false);
    assert(isUniqueBit("algorithm") == true);
    assert(isUniqueBit("hello") == false);
    assert(isUniqueBit("") == true);

    cout << "All test cases passed ✅" << endl;
    return 0;
}

File: notebooksc#/01-is-unique.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;

public class Solution01IsUnique
{
    // Solution 1: Using HashSet - O(n) time, O(n) space
    public bool IsUniqueSet(string astr)
    {
        HashSet<char> seen = new HashSet<char>();
        foreach (char c in astr)
        {
            if (!seen.Add(c))
                return false;
        }
        return true;
    }

    // Solution 2: Using bit mask - O(n) time, O(1) space (for lowercase a-z only)
    public bool IsUniqueBit(string astr)
    {
        int mark = 0;
        foreach (char c in astr)
        {
            int moveBit = c - 'a';
            if ((mark & (1 << moveBit)) != 0)
                return false;
            mark |= (1 << moveBit);
        }
        return true;
    }

    static void Main()
    {
        var sol = new Solution01IsUnique();

        // Test IsUniqueSet
        Debug.Assert(sol.IsUniqueSet("abc") == true);
        Debug.Assert(sol.IsUniqueSet("leetcode") == false);
        Debug.Assert(sol.IsUniqueSet("algorithm") == true);
        Debug.Assert(sol.IsUniqueSet("hello") == false);
        Debug.Assert(sol.IsUniqueSet("") == true);

        // Test IsUniqueBit
        Debug.Assert(sol.IsUniqueBit("abc") == true);
        Debug.Assert(sol.IsUniqueBit("leetcode") == false);
        Debug.Assert(sol.IsUniqueBit("algorithm") == true);
        Debug.Assert(sol.IsUniqueBit("hello") == false);
        Debug.Assert(sol.IsUniqueBit("") == true);

        Console.WriteLine("All test cases passed ✅");
    }
}
